providers/axfrddns/axfrddnsProvider.go

```diff
@@ -327,162 +336,284 @@ func (c *axfrddnsProvider) GetZoneRecords(domain string) (models.Records, error)
-				Msg: msg,
-				F: func() error {
-
-					// An RFC2136-compliant server must silently ignore an
-					// update that inserts a non-CNAME RRset when a CNAME RR
-					// with the same name is present in the zone (and
-					// vice-versa). Therefore we prefer to first remove records
-					// and then insert new ones.
-					//
-					// Compliant servers must also silently ignore an update
-					// that removes the last NS record of a zone. Therefore we
-					// don't want to remove all NS records before inserting a
-					// new one. For the particular case of NS record, we prefer
-					// to insert new records before ot remove old ones.
-					//
-					// This remarks does not apply for "modified" NS records, as
-					// updates are processed one-by-one.
-					//
-					// This provider does not allow modifying the TTL of an NS
-					// record in a zone that defines only one NS. That would
-					// would require removing the single NS record, before
-					// adding the new one. But who does that anyway?
-
-					update := new(dns.Msg)
-					update.SetUpdate(dc.Name + ".")
-					update.Id = uint16(c.rand.Intn(math.MaxUint16))
-					for _, c := range create {
-						if c.Desired.Type == "NS" {
-							update.Insert([]dns.RR{c.Desired.ToRR()})
-						}
-					}
-					for _, c := range del {
-						update.Remove([]dns.RR{c.Existing.ToRR()})
-					}
-					for _, c := range mod {
-						update.Remove([]dns.RR{c.Existing.ToRR()})
-						update.Insert([]dns.RR{c.Desired.ToRR()})
-					}
-					for _, c := range create {
-						if c.Desired.Type != "NS" {
-							update.Insert([]dns.RR{c.Desired.ToRR()})
-						}
-					}
+	if changes == nil {
+		return nil, nil
+	}
 
-					client := new(dns.Client)
-					client.Net = c.updateMode
-					client.Timeout = dnsTimeout
-					if c.updateKey != nil {
-						client.TsigSecret =
-							map[string]string{c.updateKey.id: c.updateKey.secret}
-						update.SetTsig(c.updateKey.id, c.updateKey.algo, 300, time.Now().Unix())
-						if c.updateKey.algo == dns.HmacMD5 {
-							client.TsigProvider = md5Provider(c.updateKey.secret)
-						}
-					}
+	hasNsUpdate := false
+	for _, change := range changes {
+		switch change.Type {
+		case diff2.CREATE:
+			if change.New[0].Type == "NS" && change.New[0].Name == "@" {
+				hasNsUpdate = true
+				continue
+			}
+		case diff2.CHANGE:
+			if change.Old[0].Type == "NS" && change.Old[0].Name == "@" {
+				hasNsUpdate = true
+				continue
+			}
+		case diff2.DELETE:
+			if change.Old[0].Type == "NS" && change.Old[0].Name == "@" {
+				hasNsUpdate = true
+				continue
+			}
+		}
+	}
 
-					msg, _, err := client.Exchange(update, c.master)
-					if err != nil {
-						return err
-					}
-					if msg.MsgHdr.Rcode != 0 {
-						return fmt.Errorf("[Error] AXFRDDNS: nameserver refused to update the zone: %s (%d)",
-							dns.RcodeToString[msg.MsgHdr.Rcode],
-							msg.MsgHdr.Rcode)
+	if hasNsUpdate {
+		update.Insert([]dns.RR{dummyNs1})
+	}
+
+	for _, change := range changes {
```

Comment:
Add a comment explaining why diff2.DELETEs are processed before everything else.

I'm curious if this is still needed. Did you try executing the changes in the order that ByRecord() produces?  It is careful to move some DELETEs to be first.  I'm curious if it could be doing a better job.

Reply:
I gave a try and indeed, moving DELETEs to be first does not seem to be required when usung `ByRecord`. At least the testsuite is OK and it looked OK when quickly tested with my local environment.

Reply:
Glad it worked!  In the future, if you do find any ordering issues please let me know.  I would like to try to fix them in diff2 before you fix them in the provider.

diff2 does this for CNAMEs. It doesn't do this for MX or other records but so far I haven't seen any providers that validate those.